node/core/dispute-coordinator/src/real/participation/mod.rs

```diff
@@ -0,0 +1,432 @@
+	}
+
+	/// Message from a worker task was received - get the outcome.
+	///
+	/// Call this function to keep participations going and to receive `ParticipationStatement`s.
+	///
+	/// This message has to be called for each received worker message, in order to make sure
+	/// enough participation processes are running at any given time.
+	///
+	/// Returns: The received `ParticipationStatement` or a fatal error, in case
+	/// something went wrong when dequeuing more requests (tasks could not be spawned).
+	pub async fn get_participation_result<Context: SubsystemContext>(
+		&mut self,
+		ctx: &mut Context,
+		msg: WorkerMessage,
+	) -> FatalResult<ParticipationStatement> {
+		let WorkerMessage(statement) = msg;
+		self.running_participations.remove(&statement.candidate_hash);
+		let recent_block = self.recent_block.expect("We never ever reset recent_block to `None` and we already received a result, so it must have been set before. qed.");
+		self.dequeue_until_capacity(ctx, recent_block.1).await?;
+		Ok(statement)
+	}
+
+	/// Process active leaves update.
+	pub async fn process_active_leaves_update<Context: SubsystemContext>(
+		&mut self,
+		ctx: &mut Context,
+		update: &ActiveLeavesUpdate,
+	) -> FatalResult<()> {
+		if let Some(activated) = &update.activated {
+			match self.recent_block {
+				None => {
+					self.recent_block = Some((activated.number, activated.hash));
+					// Work got potentially unblocked:
+					self.dequeue_until_capacity(ctx, activated.hash).await?;
+				},
+				Some((number, _)) if activated.number > number => {
+					self.recent_block = Some((activated.number, activated.hash));
+				},
+				Some(_) => {},
+			}
+		}
+		Ok(())
+	}
+
+	/// Dequeue until `MAX_PARALLEL_PARTICIPATIONS` is reached.
+	async fn dequeue_until_capacity<Context: SubsystemContext>(
+		&mut self,
+		ctx: &mut Context,
+		recent_head: Hash,
+	) -> FatalResult<()> {
+		while self.running_participations.len() < MAX_PARALLEL_PARTICIPATIONS {
+			if let Some(req) = self.queue.dequeue() {
+				self.fork_participation(ctx, req, recent_head)?;
+			} else {
+				break
+			}
+		}
+		Ok(())
+	}
+
+	/// Fork a participation task in the background.
+	fn fork_participation<Context: SubsystemContext>(
+		&mut self,
+		ctx: &mut Context,
+		req: ParticipationRequest,
+		recent_head: Hash,
+	) -> FatalResult<()> {
+		if self.running_participations.insert(req.candidate_hash().clone()) {
+			let sender = ctx.sender().clone();
+			ctx.spawn(
+				"participation-worker",
+				participate(self.worker_sender.clone(), sender, recent_head, req).boxed(),
+			)
+			.map_err(Fatal::SpawnFailed)?;
+		}
+		Ok(())
+	}
+}
+
+async fn participate(
+	mut result_sender: WorkerMessageSender,
+	mut sender: impl SubsystemSender,
+	block_hash: Hash,
+	req: ParticipationRequest,
+) {
+	// in order to validate a candidate we need to start by recovering the
+	// available data
+	let (recover_available_data_tx, recover_available_data_rx) = oneshot::channel();
```

Comment:
The participation logic should start with some checks about whether we need to participate. We shouldn't participate in any disputes that we already voted in. Or is this protected somewhere else?

Reply:
That's the same as before - checked on import. Participation will make sure though that no participation gets queued, if we are already participating at that moment. What could happen that approval voting casts a vote, after we already queued participation - in that case we would in fact duplicate work, but nothing bad should happen as on `issue_local_statement` we will only cast votes for disputes we did not yet have cast a vote for.

It seems that the risk of duplicate work in this race condition should not matter much, so I would refrain complicating things by adding cancel logic for that case.